### PR TITLE
modify storageclass settings in k8s1.23

### PR DESCRIPTION
**Pull request by MinerYang**
## What this PR does / why we need it
Fix harbor e2e test in AWS and Azure clusters without CSIDriver problems in Kubernetes v1.23
Apply a `rancher.io/local-path` storageclass for clusters lack of csi driver to dynamically provisioning for PVC usages due to original **cloud in-tree provider has been deprecated in Kubernetes v1.23**. But out-of-tree CSI drivers for EBS(AWS) and Azure Virtual Disk (Azure)haven't been implemented in tkg AWS and  AZURE clusters.
 If the corresponding CSI driver is not installed in the cluster after upgrade to this version, the PV will be in pending state thus impact to the applications.
clusters included and before tkg1.5(k8s v1.22 or before) have in-tree cloud provider plugin
vsphere clusters have own storageclass with csidriver `csi.vsphere.vmware.com`

## Expected Behavior
only when clusters without csi driver neither have a `local-path-storage` would have to install it
In our current case, AWS and Azure clusters in Kubernetes v1.23 would temporarily apply this "rancher.io/local-path" storageclass in place of the default one when creating PVC.
Harbor e2e test passed successfully.

## Which issue(s) this PR fixes

Fixes: #

## Describe testing done for PR
run harbor package 2.3.3 e2e test in following clusters

```
 cd community-edition/addons/packages/harbor/2.3.3/test 
 go get k8s.io/api/storage/v1
 go mod tidy
 CGO_ENABLED=0 go run github.com/onsi/ginkgo/ginkgo -v e2e
```

tkg1.5 Vsphere cluster in Kubernetes v1.22
tkg1.5 aws cluster in Kubernetes v1.22
tkg1.6 aws cluster in Kubernetes v1.23
tce unmangement cluster

tkg1.6 aws/azure cluster in Kubernetes 1.23 will have 2 storageclass and use 'local-path'  for PVC
```
NAME                PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
default (default)   kubernetes.io/aws-ebs   Delete          Immediate              true                   3d15h
local-path          rancher.io/local-path   Delete          WaitForFirstConsumer   false                  146m
```
```
kubectl get pvc -n goharbor
NAME                              STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
data-harbor-redis-0               Bound    pvc-2b5dfe2a-3d63-43ec-a73c-a3ee68325455   1Gi        RWO            local-path     16s
data-harbor-trivy-0               Bound    pvc-bb0bf758-13ae-44f2-aead-723f429e0277   5Gi        RWO            local-path     16s
database-data-harbor-database-0   Bound    pvc-452fce0d-5094-4ed9-a32f-6e0941b2e269   1Gi        RWO            local-path     18s
harbor-jobservice                 Bound    pvc-494b3743-0ea6-4ee6-b547-84fe8892444a   1Gi        RWO            local-path     22s
harbor-registry                   Bound    pvc-a33e387e-4ea3-4b0d-88da-773b6c1a1774   10Gi       RWO            local-path     22s
```

All other clusters  using 'default' storageclass for PVC. Take tkg1.5 aws cluster in Kubernetes 1.22 for example
```
kubectl get sc
NAME                PROVISIONER             RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
default (default)   kubernetes.io/aws-ebs   Delete          Immediate           true                   35h
```
```
kubectl get pvc -n goharbor
NAME                              STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
data-harbor-redis-0               Bound    pvc-61583862-031e-4c6c-964e-f006437089b8   1Gi        RWO            default        6m25s
data-harbor-trivy-0               Bound    pvc-a01431c5-dad0-46bb-916d-1ae8dbab4fc8   5Gi        RWO            default        6m25s
database-data-harbor-database-0   Bound    pvc-dfec647c-17e4-475e-853a-7e6612cb14d7   1Gi        RWO            default        6m26s
harbor-jobservice                 Bound    pvc-06e97098-d902-4072-aa5f-4dc5226cb6ed   1Gi        RWO            default        6m29s
harbor-registry                   Bound    pvc-a6f251e0-199d-4c2f-ad2b-87663cd8fbc5   10Gi       RWO            default        6m29s
```

## Special notes for your reviewer
[kubernetes-volume-plugin-faq-for-storage-vendors](https://github.com/kubernetes/community/blob/master/sig-storage/volume-plugin-faq.md#kubernetes-volume-plugin-faq-for-storage-vendors)

[Type of persistent-volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/)

[kubernetes-1-23-release-announcement](https://kubernetes.io/blog/2021/12/07/kubernetes-1-23-release-announcement/)

### CODE REVIEW

File: addons/packages/harbor/2.3.3/test/e2e/harbor_suite_test.go [hunk 1]
```diff
@@ -173,11 +174,34 @@ func findPackageAvailableVersion(packageName string, versionSubstr string) strin
...
+apply a "rancher.io/local-path" storageclass for clusters lack of csi driver to dynamically provisioning for PVC usages
+clusters included and before tkg1.5(k8s1.22) have in-tree cloud provider plugin
+vsphere clusters have own storageclass with csi "csi.vsphere.vmware.com"
+
+only when clusters without csi driver but with annotations "migrated_plugins" will have to install local-path-storage
+aws and azure clusters would apply this "rancher.io/local-path" sc instead of the default one
+
+return storageclass name for PVC
+*/
+func setStorageClass() string {
```
Comment (heww) [inline addons/packages/harbor/2.3.3/test/e2e/harbor_suite_test.go:187]: setStorageClass -> getStorageClass

File: addons/packages/harbor/2.3.3/test/e2e/harbor_suite_test.go [hunk 2]
```diff
@@ -173,11 +174,34 @@ func findPackageAvailableVersion(packageName string, versionSubstr string) strin
...
 	Expect(err).NotTo(HaveOccurred())
 
-	return strings.Contains(output, "(default)")
+	if (csidriver == "'<nil>'") && (strings.Contains(csiMigratedPlugins, "kubernetes.io/aws-ebs,kubernetes.io/azure-disk")) {
+		// apply a new local-path-storage
+		_, err := utils.Kubectl(nil, "apply", "-f", filepath.Join("fixtures", "local-path-storage.yaml"))
+		Expect(err).NotTo(HaveOccurred())
+		return "local-path"
+	}
+	return ""
```
Comment (heww) [inline addons/packages/harbor/2.3.3/test/e2e/harbor_suite_test.go:204]: Please change the logic to get the default storage class first, then verify it's available. If there isn't a default storage class or it isn't available, install the local path and return the local path as the storage when the local path is running.

File: addons/packages/harbor/2.3.3/test/e2e/harbor_suite_test.go [hunk 3]
```diff
@@ -173,11 +174,34 @@ func findPackageAvailableVersion(packageName string, versionSubstr string) strin
...
+
+only when clusters without csi driver but with annotations "migrated_plugins" will have to install local-path-storage
+aws and azure clusters would apply this "rancher.io/local-path" sc instead of the default one
+
+return storageclass name for PVC
+*/
+func setStorageClass() string {
+	jsonPath := `jsonpath='{.items[0].metadata.annotations}'`
+	csiMigratedPlugins, err := utils.Kubectl(nil, "get", "csinodes", "-o", jsonPath)
+	fmt.Println("csiMigratedPlugins:", csiMigratedPlugins)
```
Comment (heww) [inline addons/packages/harbor/2.3.3/test/e2e/harbor_suite_test.go:190]: remove it

File: addons/packages/harbor/2.3.3/test/e2e/harbor_suite_test.go [hunk 4]
```diff
@@ -173,11 +174,34 @@ func findPackageAvailableVersion(packageName string, versionSubstr string) strin
...
+	Expect(err).NotTo(HaveOccurred())
+
+	jsonPath = `jsonpath='{.items[0].spec.drivers}'`
+	csidriver, err := utils.Kubectl(nil, "get", "csinodes", "-o", jsonPath)
+	fmt.Println("csidriver:", csidriver)
```
Comment (heww) [inline addons/packages/harbor/2.3.3/test/e2e/harbor_suite_test.go:195]: remove it

File: addons/packages/harbor/2.3.3/test/e2e/harbor_suite_test.go [hunk 5]
```diff
@@ -173,11 +179,83 @@ func findPackageAvailableVersion(packageName string, versionSubstr string) strin
...
+	}
+
+	// do not have an available storageclass, apply a local-path-stoarge for it
+	_, err := utils.Kubectl(nil, "apply", "-f", filepath.Join("fixtures", "local-path-storage.yaml"))
+	Expect(err).NotTo(HaveOccurred())
+
+	Eventually(func(g Gomega) {
+		_, err := utils.Kubectl(nil, "apply", "-f", filepath.Join("fixtures", "local-path-storage.yaml"))
+		g.Expect(err).NotTo(HaveOccurred())
+	}, time.Second*120, time.Second*5).Should(Succeed(), fmt.Sprintln("failed to apply local-path-storage"))
```
Comment (heww) [inline addons/packages/harbor/2.3.3/test/e2e/harbor_suite_test.go:210]: change timeout to 5 minutes

File: addons/packages/harbor/2.3.3/test/e2e/harbor_suite_test.go [hunk 6]
```diff
@@ -87,17 +91,18 @@ var _ = BeforeSuite(func() {
 
 	packageComponentsNamespace = "goharbor"
 
+	//storageclass = getStorageClass()
+	storageclass, isDefault = getAvailableStorageClass()
+	if isDefault {
+		storageclass = ""
+	}
+	fmt.Println("storageclass", storageclass)
```
Comment (heww) [inline addons/packages/harbor/2.3.3/test/e2e/harbor_suite_test.go:99]: Remove it

**Review by heww — DISMISSED** [review]

**Review by heww — APPROVED** [review]

**Comment by stmcginnis** [comment]
This was merged with failing tests. Please do not merge changes until all tests are passing.

Also, for future PRs please create a PR from your fork, not by creating a branch in the main repo.

**Comment by MinerYang** [comment]
> This was merged with failing tests. Please do not merge changes until all tests are passing.
> 
> Also, for future PRs please create a PR from your fork, not by creating a branch in the main repo.

Got it and sorry for the trouble. THX a lot.

**Pull request merged**